src/Data/Json/JTable/Internal.purs

```diff
@@ -0,0 +1,218 @@
+module Data.Json.JTable.Internal
+  ( renderJTableRaw, renderRows, renderThead, renderTbody, tsToRows, sortTree
+  , Tree(..), tPath, tWidth, tHeight, tKids
+  , Cell(..), cCursor, cWidth, cHeight, cJsonPrim
+  , JPath(..), Table(..)
+  , tFromJson, tMergeArray, widthOfPrimTuple, padTree
+  , cFromJson, cMergeObj, mergeObjTuple
+  , foldJsonP, toPrim, zipWithIndex, orelse, strcmp, _nattr, _cspan, _rspan
+  ) where
+
+import Math (max)
+import Data.Either
+import Data.Maybe (Maybe(..), fromMaybe, isJust)
+import Data.Tuple
+import Data.String (joinWith, localeCompare)
+import Data.Array
+import Data.Foldable
+import qualified Data.StrMap as M
+import Data.Foldable (foldl, any, all, mconcat, elem)
+import Data.Traversable (for)
+import Control.Alt
+import Control.Apply
+import Control.Bind
+import Control.Monad
+import Control.MonadPlus
+
+import Data.Argonaut.Core
+import Data.Argonaut.JCursor
+import Data.Argonaut.Parser (jsonParser)
+import Text.Smolder.HTML (thead, tbody, tr, th, td)
+import Text.Smolder.Markup (Markup(..), Attribute(..), attribute, (!))
+
+-- path of object keys, with array indices omitted
+type JPath = [String]
+
+-- rows of cells
+type Table = [[Cell]]
+
+-- header data
+data Tree = T String JPath Number Number [Tree]
+tLabel (T l _ _ _ _) = l
+tPath (T _ p _ _ _) = p
+tWidth (T _ _ w _ _) = w
+tHeight (T _ _ _ h _) = h
+tKids (T _ _ _ _ k) = k
+
+instance showTree :: Show Tree where
+  show (T l p w h k) = 
+    joinWith " " ["(T", show l, show p, show w, show h, show k, ")"]
+
+-- cell data
+data Cell = C  JCursor Number Number JsonPrim
+cCursor (C c _ _ _) = c
+cWidth (C _ w _ _) = w
+cHeight (C _ _ h _) = h
+cJsonPrim (C _ _ _ j) = j
+
+instance showCell :: Show Cell where
+  show (C c w h j) = joinWith " " ["(C", show w, show h, show c, show j, ")"]
+
+
+foldJsonP :: forall a. (JsonPrim -> a) -> (JArray -> a) -> (JObject -> a) -> Json -> a
+foldJsonP f = 
+  foldJson (\_-> f primNull) (f <<< primBool) (f <<< primNum) (f <<< primStr)
+
+toPrim :: Json -> Maybe JsonPrim
+toPrim = foldJsonP Just (const Nothing) (const Nothing)
+
+zipWithIndex xs = zip xs (0 .. ((length xs) - 1))
+
+orelse f g x = case f x of Just v -> v 
```

Comment:
I needed something like a lazy `<|>` since it was slowing down rendering.
Fixed: added type signature, refactored.I Left the last parens in the type to highlight the grouping.
```
orelse :: forall a b. (a -> Maybe b) -> (a -> b) -> (a -> b)
orelse f g = fromMaybe <$> g <*> f 
```